icu4c/source/i18n/dtptngen.cpp

```diff
@@ -803,38 +803,63 @@ DateTimePatternGenerator::staticGetBaseSkeleton(
 
 void
 DateTimePatternGenerator::addICUPatterns(const Locale& locale, UErrorCode& status) {
-    if (U_FAILURE(status)) { return; }
-    UnicodeString dfPattern;
-    UnicodeString conflictingString;
-    DateFormat* df;
-
-    // Load with ICU patterns
-    for (int32_t i=DateFormat::kFull; i<=DateFormat::kShort; i++) {
-        DateFormat::EStyle style = static_cast<DateFormat::EStyle>(i);
-        df = DateFormat::createDateInstance(style, locale);
-        SimpleDateFormat* sdf;
-        if (df != nullptr && (sdf = dynamic_cast<SimpleDateFormat*>(df)) != nullptr) {
-            sdf->toPattern(dfPattern);
-            addPattern(dfPattern, false, conflictingString, status);
-        }
-        // TODO Maybe we should return an error when the date format isn't simple.
-        delete df;
-        if (U_FAILURE(status)) { return; }
+    if (U_FAILURE(status)) {
+        return;
+    }
+    
+    LocalUResourceBundlePointer rb(ures_open(nullptr, locale.getBaseName(), &status));
+    CharString calendarTypeToUse; // to be filled in with the type to use, if all goes well
+    getCalendarTypeToUse(locale, calendarTypeToUse, status);
 
-        df = DateFormat::createTimeInstance(style, locale);
-        if (df != nullptr && (sdf = dynamic_cast<SimpleDateFormat*>(df)) != nullptr) {
-            sdf->toPattern(dfPattern);
-            addPattern(dfPattern, false, conflictingString, status);
+    // HACK to get around the fact that the old SimpleDateFormat code (actually, Calendar::getCalendarTypeForLocale() )
+    // returns "gregorian" for ja_JP_TRADITIONAL instead of "japanese"
+    if (uprv_strcmp(locale.getBaseName(), "ja_JP_TRADITIONAL") == 0) {
+        calendarTypeToUse.clear().append("gregorian", status);
+    }
```

Comment:
Is this not a problem in Java?

Reply:
Apparently not.  The failing test in ICU4C is `IntlTestDateTimePatternGeneratorAPI::testAPI()`, specifically the block of `ja_JP_TRADITIONAL` test cases at ll. 330-349.  I found that same block of test cases at ll. 580-599 of `DateTimeGeneratorTest.java`, with completely different expected results and a little note at the top that says "This is different in C++ and Java."  That bothers me-- I'd like to understand _why_ it's different and try to make it the same-- but that feels like a whole different problem than what I'm trying to solve here.

Reply:
ticket?

Reply:
Done: https://unicode-org.atlassian.net/browse/ICU-22870